src/subcommands/create_csr.rs

```diff
@@ -0,0 +1,262 @@
+// Copyright 2021 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Creates a Certificate Signing Request (CSR) from a keypair.
+
+use crate::error::{Error, Result, ToolErrorKind};
+use crate::util::sign_message_with_policy;
+use log::error;
+use parsec_client::core::interface::operations::psa_algorithm::{
+    Algorithm, AsymmetricSignature, Hash, SignHash,
+};
+use parsec_client::core::interface::operations::psa_key_attributes::{EccFamily, Type};
+use parsec_client::BasicClient;
+use rcgen::{
+    Certificate, CertificateParams, DistinguishedName, DnType, KeyPair, RcgenError, RemoteKeyPair,
+    SignatureAlgorithm, PKCS_ECDSA_P256_SHA256, PKCS_ECDSA_P384_SHA384, PKCS_RSA_SHA256,
+    PKCS_RSA_SHA384, PKCS_RSA_SHA512,
+};
+use structopt::StructOpt;
+
+/// Creates an X509 Certificate Signing Request (CSR) from a keypair, using the signing algorithm
+/// that is associated with the key.
+///
+/// The CSR is written to the standard output in PEM format by default.
+#[derive(Debug, StructOpt)]
+pub struct CreateCsr {
+    /// The name of the key to use for signing. This must be an existing key that is accessible
+    /// to the user, and it must be a signing key (either an RSA key or an elliptic curve key).
+    ///
+    /// Elliptic curve keys must use the NIST P256 or P384 curves.
+    #[structopt(short = "k", long = "key-name")]
+    key_name: String,
+
+    /// The common name to be used within the Distinguished Name (DN) specification of
+    /// the CSR.
+    #[structopt(long = "cn")]
+    common_name: Option<String>,
+
+    /// The locality name to be used within the Distinguished Name (DN) specification of
+    /// the CSR.
+    #[structopt(long = "l")]
+    locality: Option<String>,
+
+    /// The organization name to be used within the Distinguished Name (DN) specification of
+    /// the CSR.
+    #[structopt(long = "o")]
+    organization: Option<String>,
+
+    /// The organizational unit name to be used within the Distinguished Name (DN) specification
+    /// of the CSR.
+    #[structopt(long = "ou")]
+    organizational_unit: Option<String>,
+
+    /// The state name to be used within the Distinguished Name (DN) specification of the CSR.
+    #[structopt(long = "st")]
+    state: Option<String>,
+
+    /// The country name to be used within the Distinguished Name (DN) specification of the CSR.
+    #[structopt(long = "c")]
+    country: Option<String>,
+
+    /// A Subject Alternative Name (SAN) for the domain of the CSR.
+    #[structopt(long = "san")]
+    subject_alternative_name: Option<Vec<String>>,
+}
+
+/// Short-lived structure to encapsulate the key name and the client, so that we can implement the
+/// RemoteKeyPair trait for rcgen.
+struct ParsecRemoteKeyPair {
+    key_name: String,
+    public_key_der: Vec<u8>,
+    parsec_client: BasicClient,
+    rcgen_algorithm: &'static SignatureAlgorithm,
```

Comment:
I take it we're tied down to the `static` algorithms specified in `rcgen`, so if we want to expand this list (say, to other elliptic curves) we have to make a PR against it and wait for the release?

Reply:
I was trying to think of alternative ways of handling those conversions, but I couldn't come up with anything more compact because of all the checks for different aspects of `Attributes`. Unfortunately, this sort of case-matching is indeed _really_ verbose, an unfortunate side-effect of the strict typing around those attributes.